### PR TITLE
Refactor to use popper.js

### LINKED ISSUE #26: refactor to use Popper.js
### Expected Behavior

To have tooltips wrap text to fit within the viewport

### Current Behavior

Tooltip extends text off screen

### Steps to Reproduce (for bugs)

1. https://jsfiddle.net/m3eq59rq/

screen shot 2018-03-23 at 7 48 41 am

screen shot 2018-03-23 at 7 43 04 am

### PR DESCRIPTION
**Pull request by nbond211**
(closes #26 )

### Description

Refactors `Tooltip.js` to use the [Popper.js](https://popper.js.org/index.html) library.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [ ] I have documented all new methods.
- [ ] I have written tests for all new methods/functionality.
- [ ] I have written examples for all new methods/functionality.

### CODE REVIEW

**Review by davelandry — CHANGES_REQUESTED** [review]
This is looking good @nbond211, couple of quick preliminary notes:

- [x] `.data( )` can take an array of items and produce multiple tooltips at the same time, so you'll have to do some kind of for loop when creating Poppers
- [x] let's rename the `translate( )` method to `position( )`, and allow for both an Array (ie. `[0, 10]`) or an HTMLElement. If it's the Array, construct the reference element.

And on a separate note, based on how we're using this and digging around their site, I noticed that they made a generalized library on top of Popper called [Tooltip](https://popper.js.org/tooltip-examples.html). Should we be using this instead?

**Comment by nbond211** [comment]
@davelandry I looked into [Tooltip](https://popper.js.org/tooltip-examples.html) when I was getting started and decided that it wasn't flexible enough. The `Tooltip.js` library won't let us create an html element with custom styles applied and use it as a tooltip, which is what we want to do.

File: src/Tooltip.js [hunk 1]
```diff
@@ -163,20 +164,64 @@ export default class Tooltip extends BaseClass {
...
     tooltips.exit()
       .transition(t)
-        .style(`${this._prefix}transform`, "scale(0)")
-        .remove();
+        .style("opacity", 0)
+      .remove();
+
+    for (let i = 0; i < this.data.length; i++) {
+      if (that._data[i]) {
+        const tooltip = document.getElementById(`${this._className}-${that._id(that._data[i], i)}`);
```
Comment (davelandry) [inline src/Tooltip.js]: `${this._className}` should just be `d3plus-tooltip`

File: src/Tooltip.js [hunk 2]
```diff
@@ -163,27 +164,101 @@ export default class Tooltip extends BaseClass {
...
 
     if (callback) setTimeout(callback, this._duration + 100);
 
     return this;
 
   }
 
+  /**
+   @memberof Tooltip
+   @desc If *value* is specified, sets the arrow accessor to the specified function or string and returns this generator. If *value* is not specified, returns the current arrow accessor.
```
Comment (davelandry) [inline src/Tooltip.js]: Can the description be more like: "Sets the inner HTML content of the arrow element, which by default is empty."

**Review by davelandry — CHANGES_REQUESTED** [review]

**Review by davelandry — APPROVED** [review]

File: src/Tooltip.js [hunk 3]
```diff
@@ -163,27 +170,101 @@ export default class Tooltip extends BaseClass {
...
+    for (let i = 0; i < this.data.length; i++) {
+      if (that._data[i]) {
+        const tooltip = document.getElementById(`d3plus-tooltip-${that._id(that._data[i], i)}`);
+        const arrow = document.getElementById(`d3plus-tooltip-arrow-${that._id(that._data[i], i)}`);
+        const arrowHeight = arrow.getBoundingClientRect().height;
+        arrow.style.bottom = `-${arrowHeight / 2 + 1}px`;
+
+        const arrowOffset = arrowHeight / 4;
+
+        const referenceObject = that._position() instanceof Element ? that._position() : {
```
Comment (davelandry) [inline src/Tooltip.js]: This conditional block was erroring out for me when trying to supply an accessor function to an array of 2 values. I've never seen the `Element` comparitor, maybe we should just use an Array comparitor?

```jsx
const position = that._position(that._data[i]);
const referenceObject = position instanceof Array ? {
  clientWidth: 0,
  clientHeight: 0,
  getBoundingClientRect: () => ({
    top: that._position(that._data[i])[1] - arrowOffset,
    right: that._position(that._data[i])[0] - arrowOffset,
    bottom: that._position(that._data[i])[1] - arrowOffset,
    left: that._position(that._data[i])[0] - arrowOffset,
    width: 0,
    height: 0
  })
} : position;
```

**Review by davelandry — CHANGES_REQUESTED** [review]
I'm seeing a bug when trying to draw 2 tooltips, when I use the following code the 2nd tooltip is superwide:

```js
var data = [
    {"title": "D3plus Tooltip", "body": "Check out this cool table:", "x": 100, "y": 120, "label": "Position"},
    {"title": "D3plus Tooltip #2", "body": "Check out this cool table:", "x": 200, "y": 220, "label": "Position"}
  ];

  var tips = new d3plus.Tooltip()
    .data(data)
    .thead(["Axis", d => d.label])
    .tbody([
      ["x", d => d.x],
      ["y", d => d.y]
    ])
    .render();
```

**Review by davelandry — CHANGES_REQUESTED** [review]
There's still some work to be done with positioning of the tooltip itself and the arrow. Check out this test code I've been using:

```js
var tips = new d3plus.Tooltip()
  .data([{title: "Test Tooltip"}])
  .render();

window.addEventListener("mousemove", e => {
  tips.position([e.pageX, e.pageY]);
});
```

The left and right screen extremes get funny, and can we make it flip down at the top of the screen?

**Review by davelandry — APPROVED** [review]

**Pull request merged**